src/Processors/QueryPlan/ReadFromMergeTree.cpp

```diff
@@ -1468,12 +1469,55 @@ Pipe ReadFromMergeTree::spreadMarkRangesAmongStreamsFinal(
         /// If do_not_merge_across_partitions_select_final is true and there is only one part in partition
         /// with level > 0 then we won't post-process this part, and if num_streams > 1 we
         /// can use parallel select on such parts.
+
         bool no_merging_final = do_not_merge_across_partitions_select_final &&
             std::distance(parts_to_merge_ranges[range_index], parts_to_merge_ranges[range_index + 1]) == 1 &&
-            parts_to_merge_ranges[range_index]->data_part->info.level > 0 &&
-            data.merging_params.is_deleted_column.empty() && !reader_settings.read_in_order;
+            parts_to_merge_ranges[range_index]->data_part->info.level > 0 && !reader_settings.read_in_order;
+
+        /// if do_not_merge_across_partitions_select_final = 1 and the partition contains a single part,
+        /// there is no reason to handle `is_deleted` using slow inorder reading. Faster to do it with prewhere
+        if (no_merging_final && !data.merging_params.is_deleted_column.empty())
+        {
+            if (!prewhere_info)
+                prewhere_info = std::make_shared<PrewhereInfo>();
+
+            const auto & is_deleted_input = &prewhere_info->prewhere_actions.addInput(data.merging_params.is_deleted_column, std::make_shared<DataTypeUInt8>());
+            auto not_function = FunctionFactory::instance().get("not", context);
+            ActionsDAG::NodeRawConstPtrs not_is_deleted_children = {is_deleted_input};
+
+            const auto * not_is_deleted_func = &prewhere_info->prewhere_actions.addFunction(
+                not_function,
+                std::move(not_is_deleted_children),
+                {}
+            );
+
+            auto & outputs = prewhere_info->prewhere_actions.getOutputs();
+            if (outputs.empty())
+            {
+                // no existing output — just use our filter
+                outputs.push_back(not_is_deleted_func);
+            }
+            else
+            {
+                const auto * existing_output = outputs.at(0);
+                const auto and_function = FunctionFactory::instance().get("and", context);
+                ActionsDAG::NodeRawConstPtrs and_children = { existing_output, not_is_deleted_func };
+
+                const auto * combined_filter = &prewhere_info->prewhere_actions.addFunction(
+                    and_function,
+                    std::move(and_children),
+                    {}
+                );
+
+                /// We need to be careful not to remove other outputs -- they may be there on purpose. So, we just replace the main one.
```

Comment:
That's something I remember seeing and not understanding. In which case there could be multiple outputs? Asking for education purposes

Reply:
As far as I understood -- is similar cases (maybe other situations exist, but I am not sure): we want to apply filter and at the same time preserve output. The first one is always used as filter result, others are useful in cases like this